### PR TITLE
Post: Renaming MySQL Tables With Zero Downtime

### PR DESCRIPTION
**Pull request by marczych**
I've been meaning to write a post about how we renamed a table without
any downtime for a month or two now and finally got around to it
tonight. It's still a little rough so I'd appreciate any and all
feedback!

### CODE REVIEW

File: _posts/2015-08-04-renaming-tables-with-zero-downtime.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+In fact, immediately after the `RENAME TABLE` command in this example, the query backing `bob_old_view` is left untouched so it is still referencing `alice` which doesn't exist anymore.
+This doesn't matter in particular because we end up dropping the view in the next statement, but leaving a bad view does cause `mysqldump` to fail.
+
+## Missing Data in the Information Schema
+
+Also, we discovered one minor difference between the table and the view: the primary key for the view is not available in the `information_schema`.
+Our ORM uses this information to build its model of the object so it can generate queries to interact with the underlying table.
+Fortunately, we could easily provide the information it needed in code while the view was in use so it wouldn't have to look it up in the information schema table at all.
+
+There very well be more differences between tables and views that we didn't come across.
```
Comment (davidrans) [inline _posts/2015-08-04-renaming-tables-with-zero-downtime.md]: There **may/might/could** very well...

File: _posts/2015-08-04-renaming-tables-with-zero-downtime.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+# Conclusion
+
+All of the SQL commands were nearly instantaneous and worked flawlessly when used on a fairly large table with a modest number of reads and writes.
```
Comment (davidrans) [inline _posts/2015-08-04-renaming-tables-with-zero-downtime.md]: "fairly large" and "modest" are pretty ambiguous. We should be able to approximate how many columns the table had, how hard would it be to get a rough number for the reads/writes?

+1 - Adam

**Comment by davidrans** [comment]
Thanks for writing this @marczych!

**Comment by adam000** [comment]
Other than David's 2 small comments, :+1:

File: _posts/2015-08-04-renaming-tables-with-zero-downtime.md (hunk 2 above)
Comment (marczych) [inline _posts/2015-08-04-renaming-tables-with-zero-downtime.md]: That was kind of the point. The exact numbers aren't super relevant as long as it's clear that it is an important table that we rely on being available 100% of the time. Maybe it should be rephrased to something like:

> All of the SQL commands were nearly instantaneous and worked flawlessly when used to rename our guide metadata table.

File: _posts/2015-08-04-renaming-tables-with-zero-downtime.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+In fact, immediately after the `RENAME TABLE` command in this example, the query backing `bob_old_view` is left untouched so it is still referencing `alice` which doesn't exist anymore.
+This doesn't matter in particular because we end up dropping the view in the next statement, but leaving a bad view does cause `mysqldump` to fail.
+
+## Missing Data in the Information Schema
+
+Also, we discovered one minor difference between the table and the view: the primary key for the view is not available in the `information_schema`.
+Our ORM uses this information to build its model of the object so it can generate queries to interact with the underlying table.
+Fortunately, we could easily provide the information it needed in code while the view was in use so it wouldn't have to look it up in the information schema table at all.
+
+There may be more differences between tables and views that we didn't come across so exercise caution and do thorough testing before using them in production.
```
Comment (andyg0808) [inline _posts/2015-08-04-renaming-tables-with-zero-downtime.md]: Should there be a comma after _across_?

Comment (adam000) [inline _posts/2015-08-04-renaming-tables-with-zero-downtime.md]: I thought about that too. I think it's an optional comma, but I do like the way a comma would break up the (rather long) sentence a bit.

**Comment by danielbeardsley** [comment]
I think I would go with more _tabley_ names like `people => users` or `cars => vehicles` or `employees => people`...

**Comment by danielbeardsley** [comment]
This is great!  CR :+1:

**Comment by marczych** [comment]
> I think I would go with more tabley names like people => users or cars => vehicles or employees => people...

I couldn't come up with anything at the time so I just went with `alice` and `bob`. I'll change it to `people` to `users`.

**Comment by marczych** [comment]
Thank you all for the prompt and constructive feedback! I've updated the date to today because it's looking likely that we'll publish it today. I'm happy with it other than @davidrans's point about being more specific about the size of the table that we renamed: https://github.com/iFixit/itbrokeand.ifixit.com/pull/30#discussion_r36280051.

File: _posts/2015-08-04-renaming-tables-with-zero-downtime.md (hunk 2 above)
Comment (davidrans) [inline _posts/2015-08-04-renaming-tables-with-zero-downtime.md]: I was just thinking that if I'm reading this article cause I want to do something similar, and I have a table with 1 million rows, being read hundreds of times per minute, this doesn't give really tell me if I'll have a similar experience. "Fairly large" could mean 10000 rows and "modest" could be 10 times/minute.

Comment (davidrans) [inline _posts/2015-08-04-renaming-tables-with-zero-downtime.md]: That being said, maybe the likelihood of someone having a really similar use-case isn't very high. I think I like your second suggestion better than what's currently in the post.

**Comment by marczych** [comment]
Okay, I think that I addressed all of the comments. I'd like another set of eyes on it to make sure that I didn't screw anything up in the recent commits before I publish it.

File: _posts/2015-08-06-renaming-tables-with-zero-downtime.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+We recently came across a few database tables in our schema that were begging to be renamed.
+Bringing our site down for several minutes to run the migrations and deploy the updated code was unacceptable so I started investigating to come up with a solution.
+
+In this example, we will be renaming `people` to `users`.
+
+# Using a View
+
+The first thing I came up with was using [MySQL Views] to alias the table so it would be accessible by both names (`people` and `users`).
+I came to find out that [many MySQL Views are updatable and insertable][updatable views]:
```
Comment (andyg0808) [inline _posts/2015-08-06-renaming-tables-with-zero-downtime.md]: Perhaps "I learned that"? It feels a bit wordy currently.

File: _posts/2015-08-06-renaming-tables-with-zero-downtime.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+> Some views are updatable and references to them can be used to specify tables to be updated in data change statements. That is, you can use them in statements such as UPDATE, DELETE, or INSERT to update the contents of the underlying table.
+
+This is super cool because it means that _all_ of the queries, reads and writes, can be done against the new name.
```
Comment (andyg0808) [inline _posts/2015-08-06-renaming-tables-with-zero-downtime.md]: Maybe add _both_ to make it clear that this isn't a list of ['queries', 'reads', 'writes']? "it means that _all_ of the queries, both reads and writes, can be done against the new name."

File: _posts/2015-08-06-renaming-tables-with-zero-downtime.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+Here we create a view with the new name that we want to end up with:
+
+{% highlight mysql %}
+CREATE VIEW `users` AS SELECT * FROM `people`;
+{% endhighlight %}
+
+# Updating Queries in Code
+
+Now that we have a view that acts just like a proper table, we can deploy the code that updates all queries to reference `users` instead of `people`.
```
Comment (andyg0808) [inline _posts/2015-08-06-renaming-tables-with-zero-downtime.md]: Maybe "Since the view acts just like a proper table, we can now deploy the code..."? Avoids making it sound like it's something special about our view that makes it act like a table.

**Comment by andyg0808** [comment]
I've left a bunch of really picky changes. Please ignore any you disagree with!

**Comment by marczych** [comment]
> I've left a bunch of really picky changes. Please ignore any you disagree with!

Thank you for the suggestions!

**Comment by marczych** [comment]
@iFixit/coders I'm going to publish this in a few hours so do any remaining proofreading now!

**Comment by copperwall** [comment]
That was cool!  PR :pencil2:

File: _posts/2015-08-06-renaming-tables-with-zero-downtime.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+summary: We found a way to rename MySQL tables with zero downtime by using
+         updatable views and atomic rename table commands.
+---
+
+
+Sometimes you end up with tables in your database schema that just aren't named what you want anymore.
+At first glance, renaming them without any downtime isn't feasible because updating the code that uses the table at the same instant that the table is renamed isn't easily accomplished.
+
+We recently came across a few database tables in our schema that were begging to be renamed.
+Bringing our site down for several minutes to run the migrations and deploy the updated code was unacceptable so I started investigating to come up with a solution.
```
Comment (sterlinghirsh) [inline _posts/2015-08-06-renaming-tables-with-zero-downtime.md]: comma before so.

**Comment by sterlinghirsh** [comment]
CR :zap:

**Comment by kwiens** [comment]
This is neat! I didn't know all the details, that's awesome! Ping me when it's live and I'll share it up on twitter etc. Someone should also post a hacker news link.

**Comment by adam000** [comment]
:+1:
